tests/docs/test_complete_examples.py

```diff
@@ -0,0 +1,220 @@
+            number_of_lines = len(content_lines)
+            assert number_of_lines > 0
+            yield DatasetDefinitionExample(
+                path=Path(f.name),
+                start_line=1,
+                end_line=number_of_lines,
+                content_lines=content_lines,
+            )
+
+
+def create_example_test_case_id(example: DatasetDefinitionExample) -> str:
+    """Returns a test case ID for pytest from a specific DatasetDefinitionExample."""
+    return f"{example.relative_path()} L{example.start_line}:L{example.end_line}"
+
+
+@pytest.mark.parametrize(
+    "example",
+    generate_complete_ehrql_examples(),
+    ids=create_example_test_case_id,
+)
+def test_ehrql_generate_dataset_example(
+    tmp_path: Path, example: DatasetDefinitionExample
+) -> None:
+    """Test that an ehrQL generate dataset example works."""
+    tmp_filename_base = str(uuid.uuid4())
+
+    dataset_definition_content = "\n".join(example.content_lines)
+    tmp_dataset_definition_path = tmp_path / (tmp_filename_base + ".py")
+    tmp_dataset_definition_path.write_text(dataset_definition_content)
+
+    tmp_dataset_path = tmp_path / (tmp_filename_base + ".csv")
+
+    tmp_codelist_path = tmp_path / (tmp_filename_base + "_codelist.csv")
+    tmp_codelist_path.write_text("code,category\nnot_a_real_code!,not_a_real_category!")
+
+    codelist_fn = ehrql.codelist_from_csv
+
+    def wrapped_codelist_from_csv(*args, **kwargs):
+        """Returns the result from ehrql.codelist_from_csv.
+
+        This is used to monkeypatch the real ehrql.codelist_from_csv
+        so that we can use a mock CSV, but it:
+
+        * validates the function arguments
+        * calls the real function with the mock CSV data
+
+        Different documentation examples may refer to different CSV columns.
+        Because of this, we change the arguments passed to codelist_from_csv().
+        """
+        assert (len(args) == 1) or kwargs.get(
+            "filename"
+        ) is not None, "codelist_from_csv arguments not set correctly"
+
+        try:
+            kwargs["column"] = "code"
+        except KeyError:
+            assert False, "codelist_from_csv arguments not set correctly"
+        try:
+            kwargs["category_column"] = "category"
+        except KeyError:
+            assert False, "codelist_from_csv arguments not set correctly"
+
+        return codelist_fn(filename=tmp_codelist_path, **kwargs)
+
+    with (
+        unittest.mock.patch(
+            "ehrql.codelist_from_csv",
+            wraps=wrapped_codelist_from_csv,
+        ),
+        # There is no codelist code that satisfies constraints of all code systems,
+        # so patch out the validity check and just pass in a fake codelist.
+        # If we adopted a standard for example column names, we could use valid codes.
+        unittest.mock.patch(
+            "ehrql.codes.BaseCode.__post_init__",
+            return_value=None,
+        ),
+    ):
```

Comment:
@inglesp: I've had another go at running `codelist_from_csv()` with less mocking, so the real `codelist_with_csv()` function is now at least called, albeit with patched arguments.

(I can do more to validate the filename is a correct argument.)

We always need to patch out the filename to point a temporary CSV. If we decided on a convention for these in examples: like `column="snomedct"`, we could actually pass through the `column` and `category_column`.

This would then allow actual codes to be used: we could create the temporary CSV could have these columns.

I'm not sure if this is getting too complex to be worthwhile though.

Reply:
(I fixed the checks here to use `inspect.signature`.)